### ISSUE TITLE
VERY experimental support for YTM app

### ISSUE THREAD

**Issue by th3an7** [pylerSM/YouTubeBackgroundPlayback#362]
I've created a new branch that supports YouTube Music - it works ok in video mode (playing songs as videos) - sound only is not available without YouTube RED subscription...

The .java code is exactly the same as original one for normal YT app - the only changes I've made are hooks direct URL and YTM package name...

I've tested it on Genymotion android emulator with Xposed Framework installed and it's working fine - music is played in a background with playback controls visible as notification

It's also confirmed that it's working on a physical devices - Honor 8 and Galaxy S4 - [XDA](https://forum.xda-developers.com/showpost.php?p=76874202&postcount=285)

If you want to test it yourself, check this XDA post [HERE](https://forum.xda-developers.com/showpost.php?p=76873845&postcount=283)

If anyone better at programming wants to try merging YTM and YT apps support together in one YT/MBP Xposed module - there you have it :)

**Comment by th3an7** [comment]
Forgot about one thing - if YouTube Music is not available in your country - use US VPN or setup a proxy connection for your WiFi network...

**Comment by wokawoka** [comment]
In addition, I tested your 0.1 version with an old yt music 2.23.56 version and the music unfortunately stops when the app is minimized. 
It's strange because that old ytm version doesn't even add any notification with controls on my phone, differently to what you were showing in your video. Maybe I'd need to find a 2.23.55 apk but I couldn't find online.

**Comment by wokawoka** [comment]
Thank you very much for your reply @th3an7 
I've just tried again with YTM version 2.53.54 and it looks like I'm experiencing the same problem as with version 2.23.56 (that I previously downloaded from apkmirror as well):
the music stops when the app is minimized and there is absolutely no notification with play controls like in the video I've seen online. The xposed module is correcltly activated (v.0.1-experimental4), I'm running stock android 7.1 with latest xposed installed

I'm just wondering if the fact that I'm using youtube vanced could interfer with your module. If your module is targeting the YTM specific hooks I tend to believe that it shouldn't.

Any other idea/suggestion? 
Thank you very much and keep up with the great work on YouTubeBackgroundPlayback

**Comment by wokawoka** [comment]
Thank you for replying. 
Nothing interesting in the xposed log, this is what I've found in the logcat:

```
12-11 17:41:17.456 I/YTMusicBackgroundPlayback(30213): No hook details were found for the version of YouTube installed on your device. [vc:25354330]

12-11 17:41:17.456 I/YTMusicBackgroundPlayback(30213): Using the fall-back hook details for YouTube as a last resort.

12-11 17:41:17.461 W/YTMusicBackgroundPlayback(30213): One of the hooks could not be applied: java.lang.ClassNotFoundException: com.google.android.libraries.youtube.player.background.BackgroundTransitioner

12-11 17:41:17.465 W/YTMusicBackgroundPlayback(30213): One of the hooks could not be applied: java.lang.ClassNotFoundException: com.google.android.libraries.youtube.innertube.model.PlayabilityStatusModel

12-11 17:41:17.465 W/YTMusicBackgroundPlayback(30213): Ignoring an unrecognized hook action for YouTube in the list. [a:return_string;vc:25354330]

12-11 17:41:17.469 W/YTMusicBackgroundPlayback(30213): One of the hooks could not be applied: java.lang.ClassNotFoundException: com.google.android.apps.youtube.app.background.BackgroundSettings

12-11 17:41:17.473 I/YTMusicBackgroundPlayback(30213): Done applying the hooks. [vc:25354330]
```

Any suggestion?

**Comment by th3an7** [comment]
Hmm... looks like it didn't find hooks for your version despite it's already added here on github...
`12-11 17:41:17.456 I/YTMusicBackgroundPlayback(30213): No hook details were found for the version of YouTube installed on your device. [vc:25354330]`

2.53.54 = 25354330/1000 which is OK

Try restarting your device or force closing YTMusic app once again... We'll try something...

**Comment by wokawoka** [comment]
You know what, I rebooted once again and not it works: "done applying hooks"!
It's quite strange because I already did reboot at least once previously.
Do you know why it needed more than one reboot in order to be able to correctly get the hooks?

Youtube music in the background It's so cool.
If now I could get rid of the ads (on some more recent versions they were blocked from my host file but youtube music from time to time was still interrupting the music and playing a short jingle) , I still have to test this version tough.

Do you by chance plan  to eventually release an update version working with a more recent version? It would be great, I can't understand why other users haven't commented much here or on xda (or at least that's the impression I got) because being able to reduce youtube music is even more useful than doing it on regular youtube.
Thank you once again for your prompt  assistance and for the generous job you are doing.

**Comment by th3an7** [comment]
@wokawoka I knew it... I am so f****** stupid... I was updating hooks but forgot that the link inside the module points to wrong repo 🤦‍♂️ 

Note to myself - release either update for module with corrected link OR update hooks on my fork...


About module - it's useless for most users since YT Music works only in US and to test it I had to use emulator with VPN - that's why it's experimental AF - probably only You are using it right now :P

I could check the latest version once again but in the meantime, you'd better be using this one as it's working (I guess :P)

**Comment by wokawoka** [comment]
I didn't get what was the problem about the wrong repo. Are you speaking about the dropbox link from which I downloaded the xposed module? 
According to your description it was supposed to work with youtube music version 2.23.56 and 2.23.55, while it's running perfectly with version 2.23.54 atm. Anyway, it works and I'm extremely happy now.

I have also to say that I'm writing from Europe in this right moment (Germany) and youtube music works perfectly (now in the background thanks to your module!) on both my mobile and my laptop so I guess that there are some Countries outside the US where the service works (I guess it could be easy to find them out reading online).
I bet you could make more people than just me happy (especially considering that using a "freemium" spotify on a mobile is not as easy as it use to be.

**Comment by th3an7** [comment]
@wokawoka hmm... maybe they are finally expanding YTM market...

About the link - in source code there is direct link to hooks:

https://github.com/pylerSM/YouTubeBackgroundPlayback/blob/5cb6427de03c527da0a20bb9e091bb5091725ee7/java/com/pyler/youtubebackgroundplayback/YouTubeBackgroundPlayback.java#L47

I guess after uploading that module for testing purposes, I forgot to build it again with correct link...

I'll leave it for now as it is as it's working - i just need to remember to update both repos respectively...

**Comment by wokawoka** [comment]
I see. Then I just hope that you'll not change the hooks on the github repo while we are listening successfully with this combo (xposed module + ytm 2.23.54 version works).
I cross my fingers for an updated version in the future and I'll keep monitoring this repo.
Many thanks once again!

**Comment by th3an7** [comment]
Well @wokawoka your prayers may have been heard - can you please test it for 2.61.50?

As I said - there was a lot of changes so it may or may not work - backup your current version first :)

So - Install YT Music v2.61.50 and either force close app or completely reboot your device to force download new hooks

**Comment by wokawoka** [comment]
Great news!
Alright, I've backed up the current version of the module. 
But, as a general question, if each time, after a reboot, it fetches the hooks from the repo, is there a way to just freeze it like it is (say, with the hooks for the  2.23.54 version), other than, maybe, blocking its access via the firewall?
Does it mean that, since you updated the hooks on the repo, my setup would have stopped working after a reboot? Or maybe you did keep the hooks for both versions?

Anyway I'm going to install ytm 2.61.50 and report back in a few minutes.
Thanks

**Comment by th3an7** [comment]
Everything stays the same - I am just adding next version to the already
existing ones so... if latest one is not working, just downgrade and it
will be working again - that's why it said before it couldn't find hooks
for specified version as it wasn't in .json file.

No need of blocking anything - you just need to install correct version of
the app :P

**Comment by wokawoka** [comment]
You are my hero. It looks like it's working with ytm 2.61.50 as well!

While testing the 2.23.54 version I didn't get any ads (they were blocked by the adway hosts file or the one on my router, I guess) but from time to time (not really often) I was still getting the opening/closing jingle that the app plays before and after each ad.
Do you think  that this is hardcoded in the app and there's nothing to do or it may be somehow possible to remove this jingle as well? Honestly it's not bothering me much, but removing that as well would be just the perfect setup for me to go back to listen to my music in streaming.

**Comment by wokawoka** [comment]
Thank you for the suggestion.
I installed the version you linked and I've installed it. So far it's not crashing but I'm not able to say if it's removing the ads since my host file is already blocking them. I tend to think that the short sound that announces and closes the ads is hardcoded in the app and it's not being downloaded.
If I'll be able to block it I'll surely report back.
Thank you once again. I go back listening to some music now. I did appreciate your help a lot. 
Keep up with the great work you are doing

**Comment by Untzo** [comment]
is not working anymore on version 3.03.55 of YTmusic

**Comment by th3an7** [comment]
I'll try to update it... But don't expect miracles :P
YTM app could change a lot since then

In the meantime you can use some old version that's supported or just
normal YT app with YTBP Module

**Comment by Untzo** [comment]
Thx

**Comment by th3an7** [comment]
@Untzo Check if it works again - just force close YTM app or reboot your phone (FC should be enough tho)

Seems like it didn't change much so 🤞

**Comment by Untzo** [comment]
Yes it's fully working thx for your help

**Comment by Untzo** [comment]
Plz post the new link when will be readyl

**Comment by th3an7** [comment]
@Untzo sorry it took me so long... should work again - as always - just force close YTMusic app :)

**Comment by Untzo** [comment]
Yesvi confirm it's workin for version 3.05.54

**Comment by Untzo** [comment]
YTM just got an update. And the module is not working

**Comment by th3an7** [comment]
Will do an update if I find a while today...

**Comment by th3an7** [comment]
@Untzo I've added 3.07.54 - should work for that version
I've skipped 3.07.52 as it looks like it's only for one CPU arch and 3.07.54 is the latest one - let me know if I should still add .52 :D

**Comment by Untzo** [comment]
Thx. Fully working again. But im still thinking tha lt u should add an option to modify audio quality

**Comment by th3an7** [comment]
Like i was saying before - I was trying to add audio only option but I
failed - I think I found the class with that but when I added it into
hooks, the module all of the sudden stopped working... Nothing worked, not
even background playback so I left it as it is now...

It's just modified normal YTBP module with changed hooks and YTM package
name - the rest is exactly as it was before...

>

**Comment by Untzo** [comment]
Oooh ok understood
________________________________

**Comment by SammyTRK** [comment]
Is it just me or is this not working on Pie with EdXposed?

**Comment by Untzo** [comment]
Im on nougat and it's fully working

**Comment by th3an7** [comment]
@SammyTRK can you check logcat and xposed logs?
I didn't know there is something like EdXposed... looking at their readme on github, I can see [Known Issues](https://github.com/ElderDrivers/EdXposed#known-issues) - especially
* resources hooking is not supported yet
* only a few Xposed modules has been tested for working

If that's the case, I don't think it will work - try using official Xposed for Pie (if there is one... I haven't been watching for updates since I didn't use Xposed Framework for a while now)

**Comment by SammyTRK** [comment]
I have a Note5(Android 6.0.1) and a Note9(Oreo8.1 or Pie9.0). I have tried every version of this module I still have with the matching version of YTmusic on every system and nothing works for me anymore. I even restored and old backup on my note5 it has version 2.61.50 on it which I know was working back then and it doesn't work now. I have no idea what is going on. I can give logs for official Xposed for Oreo and then for EdXposed on Pie. It might be late tonight or tommorrow. 

If someone would send me a known working version of this module that would cut out a variable. I would think that if the build is successful in AS it should be a working module right? Up until now every version I have built has worked in the past. Now none of them work. 

Also, I wanted to ask about the latest update. There is an extra bracket on line 163, I removed it in AS before building should I keep it looking like this?
            ] }
        ]],
        "30754": [

**Comment by th3an7** [comment]
@SammyTRK good catch - it shouldn't be there...
Also if you're building it yourself and you're using your own hooks, you need to change direct url in:
```java
private static final String HOOKS_DOWNLOAD_URL = "https://raw.githubusercontent.com/pylerSM/YouTubeBackgroundPlayback/yt-music/assets/hooks-music.json";
```

But if you're just building .apk and not updating hooks yourself - it should be ok...
I've also just remembered, pre-built module .apk has different URL which directs to hooks in my fork - if you've built module from this source - it should work again as I've deleted the extra `]`

Try restarting your device or force close YTM app and post the result

**Comment by SammyTRK** [comment]
OMG it is working again. Version 3.07.54 on Note9 Oreo 8.1. I'll install Pie again later and edit this post to let you know if its working with EdXposed. Thanks man!

**Comment by th3an7** [comment]
@SammyTRK it should work if it works for Oreo - I've fixed the hooks by deleting the extra ']' as you pointed out :D
Sorry I didn't noticed it while I was updating them :P

**Comment by th3an7** [comment]
@Untzo there is no new YTM app on apkmirror - can you upload it somewhere so I can quickly update hooks?

**Comment by Untzo** [comment]
nice bo thx so much

**Comment by wokawoka** [comment]
are there any benefits in uptdating the ytm app each time they release a new version? (and therefore having to request each time to th3an7 an updateof the hooks?)

I'm personally listening to music on an old version of a couple of months ago and it works perfectly.  did they add new functions in the newer versions?
Maybe the possibility to select and use the "music only" switch would possibly may be a good reason to switch for me. But for now, it if it works, I'll not change it
thank you th3an7

**Comment by th3an7** [comment]
Taking about "music only" option - does someone have latest apk of YTM (v3.09.57) but for x86 CPU arch? Can't find that anywhere and I'd like to do some tests on emulator to check if I am actually retarded and can't find correct hooks, or if it's not possible at all...

**Comment by SammyTRK** [comment]
Will this work? First one I found It may be modded. It says Cpu: x86
http://youtube-music.apk.black/

**Comment by wokawoka** [comment]
If that apk doesn't work I have an android x86 tablet at home so I can eventually install it and extract the apk when needed. Just let me know

**Comment by th3an7** [comment]
Nah, it works... I couldn't set up VPN on that so the app would launch so I
am doing tests now on my own personal device

Thanks :D

**Comment by wokawoka** [comment]
alright, great, happy tests then

**Comment by Untzo** [comment]
Hey yt muwic got and update

**Comment by th3an7** [comment]
Will try updating it later if that's ok...
Just wanna say it's 6:40 am and I still didn't go to sleep and I've just
turned off PC

Expect the update in max 12h...

**Comment by Untzo** [comment]
It's perfectly fine. Thanks
________________________________

**Comment by th3an7** [comment]
@Untzo I've added support for 3.11.54 - there were some changes in YTM app code so let me know if it works or not :P

**Comment by Untzo** [comment]
Works like a charm. Thank you so much
________________________________

**Comment by Untzo** [comment]
Yt music got updated. Can you make a magisk module?

**Comment by th3an7** [comment]
Yeah, will check it out in a while - should be up and working again soon
(probably 2h max)

**Comment by Untzo** [comment]
Can you something for magisk too?

**Comment by th3an7** [comment]
Not really... It's just YTBP module with changed hooks and package name -
rest of the code is unchanged.
When I was checking YTM, I didn't even think it was possible so...

If anyone else wants to try port Xposed module to magisk - go ahead, I
probably don't even know what to do first :P

**Comment by SammyTRK** [comment]
I just used your [old guide](https://www.youtube.com/watch?v=OG8fvxnZnSo) to try and update for the new ytmusic, and it actually worked. I have no idea what I was looking at or how you find them to begin with but the files matched in looks so I tried it and it works!
I saved you a little time with this update. :) Here is the updated hooks:
```
         ],
        "31758": [
            { "class_name": "vym", "method_name": "k", "parameter_types": [], "actions": [
                { "name": "set_field_boolean_before_method", "field_name": "q.b", "value": true }
            ] },
            { "class_name": "wvc", "method_name": "d", "parameter_types": [ "acsp" ], "actions": [
                { "name": "return_boolean", "value": true }
            ] }
```

**Comment by th3an7** [comment]
Thanks @SammyTRK - I've added your hooks so it should work again

BTW: if you're using IDA (or some other decompiler) you can use "Search for sequence of bytes (ALT+B) and search for:
First class - `55 ?? ?? ?? 39 ?? ?? ?? 54 ?? ?? ?? 72 ?? ?? ?? ?? ?? 0C ?? 1F ?? ?? ?? 60`
Second class - `12 00 39 03 03 00` (should be the one with method .d - so it should look like <somename>.d

If the code changes you'd have to find some other HEX but for now these ones work fine - then you can just propose commit so it would go even faster :D

**Comment by SammyTRK** [comment]
Ok, thanks. I'll try it next update.

**Comment by SammyTRK** [comment]
I am curious about this part "field_name": "q.b" in the hooks file. Where do you find this? If it's not to hard to explain...

**Comment by th3an7** [comment]
It's this part - `q.b` are the blue marked letters - those 2 lines are at the beginning of the first class you're looking for

![image](https://user-images.githubusercontent.com/13398380/60349108-3812ba00-99c1-11e9-9b40-c7f7a8f09af2.png)

**Comment by SammyTRK** [comment]
Something has changed in the latest YT music app 3.33.51. I can't find the first class with that search string now. The second one is still there though.

**Comment by SammyTRK** [comment]
@th3an7 Something has changed in the latest YT music app 3.33.51. I can't find the first class with that search string now. The second one is still there though.

**Comment by th3an7** [comment]
I'll check it out
Thanks for notifying

**Comment by th3an7** [comment]
@SammyTRK I've pushed one commit - check if it works now

HEX has changed quite bit - I am guessing (as I don't know if the new one I've put is the correct one) that new class looks like the one in YT app - at least it looks similar.

Let me know if it works - if not I'll try to find something else
Used hex - `55 ?? ?? ?? 12 ?? ?? ??  3F ?? ?? ?? C2 ?? ?? ??` - it showed me only one class so it may not work for next version (if searching that in next version will show nothing, let me know)

**Comment by SammyTRK** [comment]
> @SammyTRK I've pushed one commit - check if it works now
> 
> HEX has changed quite bit - I am guessing (as I don't know if the new one I've put is the correct one) that new class looks like the one in YT app - at least it looks similar.
> 
> Let me know if it works - if not I'll try to find something else
> Used hex - `55 ?? ?? ?? 12 ?? ?? ?? 3F ?? ?? ?? C2 ?? ?? ??` - it showed me only one class so it may not work for next version (if searching that in next version will show nothing, let me know)

Yes. That works. Thanks.

**Comment by SammyTRK** [comment]
@th3an7 The new version of YT music v3.53.51 has changed the second method somewhat. I can find it with the same search but the whole method has changed some. I tried updating it as I usually do but its not working. :(

Also, I found LuckyPatcher has a custom patch for YTM that gives background playback, gets rid of ads and enables audio only. You should check it out maybe you can use that to enable those features in this app. Although I'm guessing it probably won't work with 3.53.51 either (haven't tried it yet).

**Comment by th3an7** [comment]
> @th3an7 The new version of YT music v3.53.51 has changed the second method somewhat. I can find it with the same search but the whole method has changed some. I tried updating it as I usually do but its not working. :(

What were the classes you found? It may be the issue with first one actually as I can't seem to find it myself (just did quick search with old HEXes...)
Second one may be `vog.c` `adpe` (at least that's what LP patch looks for...)

> Also, I found LuckyPatcher has a custom patch for YTM that gives background playback, gets rid of ads and enables audio only. You should check it out maybe you can use that to enable those features in this app. Although I'm guessing it probably won't work with 3.53.51 either (haven't tried it yet).

Ye, there are also patches for YT (that's how I was looking for classes to update the module before :D) Second class may be working (see above) but dunno what about the rest (and if Xposed can hook them properly)...

If you don't mind, you could DM me on Discord - `Th3Ant (aka ✪ MERITT)#9411` so it would be easier to talk (or somewhere else... just send me info where we could figure out something quicker :D (might not respond immediately so I'll check when I can - probably due to timezones)

**Comment by SammyTRK** [comment]
@th3an7 Ok.

**Comment by Sajito** [comment]
@th3an7 @SammyTRK have you two figured out what changed? Am running YTM 3.61.51 without any hooks present.

**Comment by th3an7** [comment]
> @th3an7 @SammyTRK have you two figured out what changed? Am running YTM 3.61.51 without any hooks present.

I've pushed update with new hooks - it should work (if not, let me know :P )

**Comment by someone5678** [comment]
https://apkcombo.com/youtube-music-stream-songs-music-videos/com.google.android.apps.youtube.music/download/apk

3.69.52 is out and module does'nt work

**Comment by th3an7** [comment]
> https://apkcombo.com/youtube-music-stream-songs-music-videos/com.google.android.apps.youtube.music/download/apk
> 
> 3.69.52 is out and module does'nt work

Seems like there was no changes - just version bump...
Updated and should be up again

**Comment by someone5678** [comment]
https://apkcombo.com/youtube-music-stream-songs-music-videos/com.google.android.apps.youtube.music/download/apk

YTM 3.71.54 is out now!

**Comment by th3an7** [comment]
> https://apkcombo.com/youtube-music-stream-songs-music-videos/com.google.android.apps.youtube.music/download/apk
> 
> YTM 3.71.54 is out now!

Updated 😃

**Comment by someone5678** [comment]
https://apkcombo.com/youtube-music-stream-songs-music-videos/com.google.android.apps.youtube.music/download/apk

YTM 3.71.57 is out!

**Comment by someone5678** [comment]
https://apkcombo.com/en-pl/youtube-music-stream-songs-music-videos/com.google.android.apps.youtube.music/download/apk

YTM 3.75.50 has released!

**Comment by someone5678** [comment]
YTM 3.77.52 released!
https://apkcombo.com/youtube-music-stream-songs-music-videos/com.google.android.apps.youtube.music/download/apk

**Comment by th3an7** [comment]
> YTM 3.77.52 released!
> https://apkcombo.com/youtube-music-stream-songs-music-videos/com.google.android.apps.youtube.music/download/apk

It was already added yesterday I think (in commit #413 ) - is it not working? It was PR-ed so I didn't check classes myself...

**Comment by someone5678** [comment]
> It was already added yesterday I think (in commit #413 ) - is it not working? It was PR-ed so I didn't check classes myself...
Oops sorry I didn't check that 😬

**Comment by someone5678** [comment]
YTM 3.79.50 is out!

https://apkcombo.com/youtube-music-stream-songs-music-videos/com.google.android.apps.youtube.music/download/apk

**Comment by th3an7** [comment]
Sorry for the delay - should be up and working again :D

**Comment by th3an7** [comment]
@Apologize327 Moving convo here as mentioned in the other issue...
I've updated hooks for 3.81.52 - they've changed something and I may need to add new class... testing old method first

Let me know if it still doesn't work

**Comment by Apologize327** [comment]
YTM 3.83.55 is out
https://apkcombo.com/youtube-music-stream-songs-music-videos/com.google.android.apps.youtube.music/download/apk

**Comment by th3an7** [comment]
> YTM 3.83.55 is out
> https://apkcombo.com/youtube-music-stream-songs-music-videos/com.google.android.apps.youtube.music/download/apk

Hmm... I've updated hooks for that version 2 days ago.. Is it not working? Can you please check Xposed logs?
There was couple changes already in previous version but I didn't change hooks to reflect that - this one (3.83.55) is using the same method.

If it doesn't work, I'll try adding new class to hooks and see if that fixes it (or breaks even more...) - but yeah, check Xposed logs if there is anything useful for now

**Comment by Apologize327** [comment]
Sorry ,my bad, it's working, just need re-installing YTM but not update directly every time to make it work.

**Comment by th3an7** [comment]
> Sorry ,my bad, it's working, just need re-installing YTM but not update directly every time to make it work.

Hmm... that doesn't sound right - it should work after updating too...
If it happens again, can you check Xposed logs anyway and/or if it's possible, could you send `base.apk` from `data/app/` (if it exists)? 

Wonder why it didn't apply hooks...

**Comment by Apologize327** [comment]
> > Sorry ,my bad, it's working, just need re-installing YTM but not update directly every time to make it work.
> 
> Hmm... that doesn't sound right - it should work after updating too...
> If it happens again, can you check Xposed logs anyway and/or if it's possible, could you send `base.apk` from `data/app/` (if it exists)?
> 
> Wonder why it didn't apply hooks...

ok, I'll check it next version

**Comment by someone5678** [comment]
YTM 3.89.52 released!

https://apkcombo.com/youtube-music-stream-songs-music-videos/com.google.android.apps.youtube.music/

**Comment by th3an7** [comment]
> YTM 3.89.52 released!
> 
> https://apkcombo.com/youtube-music-stream-songs-music-videos/com.google.android.apps.youtube.music/

Updated - some things have changed so let me know if it still works (hope the script still shows correct classes 🤞

**Comment by someone5678** [comment]
> > YTM 3.89.52 released!
> > https://apkcombo.com/youtube-music-stream-songs-music-videos/com.google.android.apps.youtube.music/
> 
> Updated - some things have changed so let me know if it still works (hope the script still shows correct classes 🤞

Working without any problem Thx!!!
